ADDON-AUTHOR-GUIDE.md

```diff
@@ -88,3 +88,130 @@ To avoid these limitations, you will want to have a build step that allows you t
 - When automatically upgrading addons to **v2**, `@embroider/compat` already has the full application available to examine, so any configuration-dependent transformations will run with the right configuration. For example, any polyfills that behave differently under different Ember versions will see the app's Ember version and do the right thing. But when publishing a native **v2** package, you _don't_ have that information available so if you want to emit different code for different Ember versions the polyfill needs to use `@embroider/macros`.
 
 While it's important that we ship good shared tooling, a nice thing about having a stable **v2** format standard is that the community is free to experiment with alternative addon build tooling, and it won't impact compatibility or stability.
+
+## Replacing the {{component}} helper
+
+Using the `{{component}}` helper (with one small exception which we'll get to) prevents an addon from achieving the **Optimized Embroider Safe** support level. So addons should stop using it. But how?
```

Comment:
I believe this is not *quite* true, i.e. there are at least two valid exceptions: the one you mention (static string), but also when yielding contextual components you will still need the `{{component}}` helper to curry arguments. And this will still work with static components, as long as the component positional argument itself is guaranteed to refer to a component that embroider was able to "see" during its analysis. 

Example: https://github.com/kaliber5/ember-bootstrap/pull/1251/files#diff-c3c2155cba3ac783f3443fa7014f24ced9dac371164096f02cb9d1c0767a651eR8-R9. 

Admittedly there is a loophole here as `@titleComponent` could be a string, but I guess I could now wrap that with the `ensureSafeComponent` *helper*, right? Anyways, as the example shows you still need the `component` helper to curry arguments. FWIW, I have that pattern quite a lot in ember-bootstrap, where contextual components are yielded, but they can *also* be passed as arguments to exchange the used component class.

Reply:
This is a good point. We should expand this section of the guide to help people with those cases. While there are plenty of patterns that are safe, there really is only one that embroider can *see* is safe, and that is a string literal argument.

We need to think about what the future of currying looks like in Ember, because I suspect it doesn't need to continue being a special feature of the `{{component}}` helper.

Once we're free to pass around component classes and then invoke them, "currying" can just be `class extends`.

Also, I think named blocks replaces many cases where you would have passed a curried component as an argument:

```hbs
<Menu @titleBar={{component "fancy-bar" mode=this.mode}} />
{{!-- vs --}}
<Menu>
  <:titleBar>
    <FancyBar @mode={{this.mode}} />
  </:titleBar>
</Menu>
```

Notice that this also solves the problem of "how do I curry HTML attributes?".

Reply:
Your comment makes me realize I do need to explain the rule more precisely.

You can use the component helper, but only if the first argument is: 
 - a string literal
 - or a call to the `ensure-safe-component` helper.

I didn't say so explicitly, but embroider should be made to not worry about this pattern:

```hbs
{{component (ensure-safe-component @whatever) }}
```

I didn't mention that case because I was assuming it would never be needed (use angle brackets instead), but you're right that if you want to curry arguments onto it but not invoke it, you'd need that pattern.

Reply:
This is now [implemented](https://github.com/embroider-build/embroider/pull/608) and [documented](https://github.com/embroider-build/embroider/pull/611)